### ISSUE TITLE
R-012: Embedding service (sentence-transformers)

### ISSUE THREAD

**Issue by nhlongnguyen** [nhlongnguyen/mindbridge#12]
## Description
Implement comprehensive embedding service using sentence-transformers to generate high-quality vector embeddings for both code documentation and structured content. This service provides the semantic understanding foundation for the RAG system, converting text content into dense vector representations optimized for similarity search and retrieval.

### Scenario 1: Documentation Content Embedding
**As a developer**, I want documentation content to be converted to embeddings so that I can perform semantic search across technical documentation using natural language queries.

*Example*: When processing FastAPI documentation chunks:
- **Input text**: "FastAPI automatically generates OpenAPI documentation for your API endpoints. You can customize the documentation by adding descriptions, examples, and response models to your route functions."
- **Output**: 1536-dimensional vector embedding capturing semantic meaning
- **Usage**: Enables queries like "how to customize API docs" to find this content even without exact keyword matches
- **Context preservation**: Related chunks (code examples, implementation details) have similar embeddings

### Scenario 2: Code Structure Embedding  
**As a developer**, I want code components (functions, classes, docstrings) to be embedded so that I can find related functionality through semantic similarity rather than just text matching.

*Example*: When processing code documentation:
```python
def authenticate_user(username: str, password: str) -> Optional[User]:
    """Authenticate user credentials and return user object.
    
    Validates username and password against the database,
    handles password hashing verification, and returns
    the authenticated user or None if invalid.
    """
```
The embedding should capture:
- **Functional concept**: Authentication, user validation
- **Technical context**: Database interaction, password security
- **Return behavior**: User object or None
- **Semantic relationships**: Links to related auth functions, user management

### Scenario 3: Batch Processing with Contextual Optimization
**As a developer**, I want efficient batch processing of large documentation sets so that embedding generation is optimized for performance while maintaining quality.

*Example*: When processing a complete repository documentation:
- **Input**: 500+ documentation chunks from crawled content and processed Markdown
- **Processing**: Batch embeddings in groups of 32-64 for optimal GPU/CPU utilization
- **Context preservation**: Related chunks maintain semantic proximity in vector space
- **Quality assurance**: Embedding quality metrics validate semantic coherence
- **Output**: Structured embeddings ready for Neo4j vector storage

### Core Embedding Generation
- [ ] Generate 1536-dimensional embeddings using sentence-transformers
- [ ] Support multiple pre-trained models (all-MiniLM-L6-v2, all-mpnet-base-v2)
- [ ] Handle variable-length text inputs (50-512 tokens optimal range)
- [ ] Normalize embeddings for consistent similarity calculations
- [ ] Generate embeddings for both documentation and code content

### Batch Processing Optimization
- [ ] Efficient batch processing with configurable batch sizes
- [ ] GPU acceleration when available, CPU fallback when not
- [ ] Memory-efficient processing for large document sets
- [ ] Progress tracking and estimation for long-running operations
- [ ] Concurrent processing with proper resource management

### Content-Aware Processing
- [ ] Preprocessing optimization for technical documentation
- [ ] Handle code snippets and technical terminology effectively
- [ ] Preserve semantic meaning across different content types
- [ ] Context-aware chunking integration for optimal embedding quality
- [ ] Special handling for structured content (tables, lists, code blocks)

### Quality Assurance
- [ ] Embedding quality validation and consistency checks
- [ ] Similarity score validation for related content
- [ ] Outlier detection for potentially problematic embeddings
- [ ] Embedding dimension and format validation
- [ ] Performance metrics and quality reporting

### Integration and Storage
- [ ] Integration with Neo4j for vector storage
- [ ] Metadata association with embeddings (source, type, context)
- [ ] Efficient serialization and deserialization
- [ ] Caching mechanism for duplicate content
- [ ] Error handling and retry logic for failed generations

### Performance and Scalability  
- [ ] Embedding generation rate of 100+ documents per minute
- [ ] Memory usage optimization for large batches
- [ ] Resource monitoring and management
- [ ] Graceful degradation under resource constraints
- [ ] Benchmarking and performance profiling capabilities

### Core Embedding Service
```python
import torch
from sentence_transformers import SentenceTransformer
import numpy as np
from typing import List, Dict, Any, Optional, Union, Tuple
from dataclasses import dataclass
import asyncio
from concurrent.futures import ThreadPoolExecutor
import logging
from pathlib import Path

@dataclass 
class EmbeddingRequest:
    text: str
    content_type: str  # 'documentation', 'code', 'comment', 'title'
    metadata: Dict[str, Any]
    source_id: Optional[str] = None
    context: Optional[str] = None

@dataclass
class EmbeddingResult:
    embedding: List[float]
    text: str
    content_type: str
    metadata: Dict[str, Any]
    processing_time: float
    model_name: str
    quality_score: Optional[float] = None

class SentenceTransformerService:
    def __init__(
        self, 
        model_name: str = "sentence-transformers/all-MiniLM-L6-v2",
        device: Optional[str] = None,
        batch_size: int = 32
    ):
        self.model_name = model_name
        self.device = device or ('cuda' if torch.cuda.is_available() else 'cpu')
        self.batch_size = batch_size
        self.model: Optional[SentenceTransformer] = None
        self.executor = ThreadPoolExecutor(max_workers=4)
        self.logger = logging.getLogger(__name__)
        
    async def initialize(self) -> None:
        """Initialize the sentence transformer model"""
        
    async def generate_embeddings(
        self, 
        requests: List[EmbeddingRequest]
    ) -> List[EmbeddingResult]:
        """Generate embeddings for batch of requests"""
        
    async def generate_single_embedding(
        self, 
        text: str, 
        content_type: str = "documentation"
    ) -> List[float]:
        """Generate embedding for single text input"""
        
    def _preprocess_text(self, text: str, content_type: str) -> str:
        """Preprocess text based on content type"""
        
    def _validate_embedding(self, embedding: np.ndarray) -> bool:
        """Validate embedding quality and format"""
```

## Dependencies
- **Requires**: R-027 (Contextual RAG chunking) - Provides optimized text chunks for embedding
- **Requires**: R-011 (Markdown document processor) - Provides processed documentation content
- **Enables**: R-025 (Neo4j database integration) - Provides embeddings for vector storage
- **Enables**: R-015 (Vector similarity search) - Provides embeddings for search functionality
- **Integrates with**: R-019 (Error handling) - Uses error handling framework

## Priority & Complexity
- **Priority**: Critical - Core component for semantic search and RAG functionality
- **Complexity**: High - ML model integration, batch optimization, quality assurance
- **Size**: Large (12-14 hours)
- **Phase**: Async Processing
- **Estimated Days**: 1.5

## Success Metrics
- [ ] Generate embeddings at 100+ documents per minute rate
- [ ] Maintain embedding quality score above 0.85
- [ ] Handle batches of 1000+ documents without memory issues
- [ ] Achieve 95%+ similarity coherence for related content
- [ ] Support both CPU and GPU processing efficiently
